SeQuant/core/export/python_einsum.hpp

```diff
@@ -0,0 +1,636 @@
+#ifndef SEQUANT_CORE_EXPORT_PYTHON_EINSUM_HPP
+#define SEQUANT_CORE_EXPORT_PYTHON_EINSUM_HPP
+
+#include <SeQuant/core/export/context.hpp>
+#include <SeQuant/core/export/generator.hpp>
+#include <SeQuant/core/expr.hpp>
+#include <SeQuant/core/index.hpp>
+#include <SeQuant/core/space.hpp>
+#include <SeQuant/core/utility/macros.hpp>
+#include <SeQuant/core/utility/string.hpp>
+
+#include <range/v3/view/join.hpp>
+#include <range/v3/view/transform.hpp>
+
+#include <boost/unordered/unordered_map.hpp>
+#include <boost/unordered/unordered_set.hpp>
+
+#include <cctype>
+#include <cstdlib>
+#include <sstream>
+#include <string>
+#include <string_view>
+#include <vector>
+
+namespace sequant {
+
+/// Backend selection for Python einsum generator
+enum class PythonEinsumBackend {
+  NumPy,   ///< Generate code using numpy.einsum
+  PyTorch  ///< Generate code using torch.einsum
+};
+
+/// Context class for the PythonEinsumGenerator
+class PythonEinsumGeneratorContext : public ExportContext {
+ public:
+  using ShapeMap = std::map<IndexSpace, std::string>;
+
+  PythonEinsumGeneratorContext(
+      PythonEinsumBackend backend = PythonEinsumBackend::NumPy)
+      : m_backend(backend) {}
+  ~PythonEinsumGeneratorContext() = default;
+  PythonEinsumGeneratorContext(
+      ShapeMap index_shapes,
+      PythonEinsumBackend backend = PythonEinsumBackend::NumPy)
+      : m_index_shapes(std::move(index_shapes)), m_backend(backend) {}
+
+  /// Get the dimension/shape for a given index space
+  std::string get_shape(const IndexSpace &space) const {
+    auto it = m_index_shapes.find(space);
+    if (it != m_index_shapes.end()) {
+      return it->second;
+    }
+    // Default fallback - use a simple naming scheme
+    return "dim_" + toUtf8(space.base_key());
+  }
+
+  /// Get the shape tuple for a tensor
+  std::string get_shape_tuple(const Tensor &tensor) const {
+    std::string shape = "(";
+    bool first = true;
+    for (const Index &idx : tensor.const_indices()) {
+      if (!first) shape += ", ";
+      shape += get_shape(idx.space());
+      first = false;
+    }
+    shape += ")";
+    return shape;
+  }
+
+  /// Set the dimension/shape for a given index space
+  void set_shape(const IndexSpace &space, std::string shape) {
+    m_index_shapes[space] = std::move(shape);
+  }
+
+  /// Get the backend being used
+  PythonEinsumBackend backend() const { return m_backend; }
+
+  /// Set the backend
+  void set_backend(PythonEinsumBackend backend) { m_backend = backend; }
+
+  /// Get the module prefix (np. or torch.)
+  std::string module_prefix() const {
+    return m_backend == PythonEinsumBackend::NumPy ? "np." : "torch.";
+  }
+
+ protected:
+  ShapeMap m_index_shapes;
+  PythonEinsumBackend m_backend;
```

Comment:
In my head, the information what kind of code is to be generated should be a property of the generator, not the context. The conceptual separation between context and generator is more or less this: the generator determines what and how code is generated whereas the context is used for all information the generator can't know at compile-time. Of course, the lines are somewhat blurry.

I think I would even go so far to have two separate generator (and context) classes for the two different Einsum variants, where one is the base implementation and the other inherits from that and then only implements the stuff that needs to be done differently. That is also how it was done for the Julia generators.

Reply:
[d768e3c](https://github.com/ValeevGroup/SeQuant/pull/429/commits/d768e3c5c2632d6ae17a24353e955dcaac6b7dc4)